Mathlib/Data/Multiset/Order.lean

```diff
@@ -0,0 +1,513 @@
+/-
+Copyright (c) 2024 Haitian Wang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Haitian Wang, Malvin Gattinger
+-/
+import Mathlib.Tactic.Linarith
+import Mathlib.Data.Multiset.Basic
+import Mathlib.Logic.Relation
+
+/-!
+# Dershowitz-Manna ordering
+
+In this file we define the _Dershowitz-Manna ordering_ on multisets.
+We prove that, given a well-founded partial order on the underlying set,
+the Dershowitz-Manna ordering defined over multisets is also well-founded.
+
+## Main results
+
+- `MultisetLT` :  the standard definition
+- `dm_wf` :       the main theorem about the `Dershowitz-Manna ordering`.
+- `Lt_LT_equiv` : two definitions of the Dershowitz-Manna ordering are equivalent.
+
+## References
+
+* [Wikipedia, Dershowitz–Manna ordering*]
+(https://en.wikipedia.org/wiki/Dershowitz%E2%80%93Manna_ordering)
+
+* [CoLoR](https://github.com/fblanqui/color), a Coq library on rewriting theory and termination.
+  Our code here is inspired by their version of called `mOrd_wf` in the file
+  [MultisetList.v](https://github.com/fblanqui/color/blob/1.8.5/Util/Multiset/MultisetList.v).
+
+-/
+
+variable {α : Type*}
+
+/-- The standard Dershowitz–Manna ordering: -/
+inductive MultisetLT [DecidableEq α] [Preorder α] (M N : Multiset α) : Prop :=
+  | MLT : ∀ (X Y Z: Multiset α),
+    Y ≠ ∅ →
+    M = Z + X →
+    N = Z + Y →
+    (∀ x ∈ X, ∃ y ∈ Y, x < y) → MultisetLT M N
+
+/-- MultisetRedLt is a special case of MultisetLT. The transitive closure of it is used to define
+an equivalent (proved later) version of the ordering. -/
+inductive MultisetRedLt [DecidableEq α][LT α] (M N : Multiset α) : Prop :=
+  | RedLt : ∀ (X Y:Multiset α) (a : α) ,
+    (M = X + Y) →
+    (N = X + {a}) →
+    (∀ y, y ∈ Y → y < a) → MultisetRedLt M N
+
+open Relation
+
+/-- MultisetLt is the transitive closure of MultisetRedLt and is equivalent to MultisetLT
+    (proved later). -/
+def MultisetLt [DecidableEq α][LT α] : Multiset α → Multiset α → Prop := TransGen MultisetRedLt
+
+/-- AccM_1 defines the accessibility relation given MultisetRedLt. -/
+def AccM_1 [DecidableEq α][Preorder α] : Multiset α → Prop := Acc MultisetRedLt
+
+/- MultisetRedLt is a special case of MultisetLT. -/
+theorem redLt_LT [DecidableEq α] [Preorder α] (M N : Multiset α) :
+    MultisetRedLt M N → MultisetLT M N := by
+  intro hyp
+  rcases hyp with ⟨X, Y, a, M_def, N_def, ys_lt_a⟩
+  apply MultisetLT.MLT Y {a} X _ M_def N_def
+  · simpa
+  · simp
+
+/- Some useful lemmas about Multisets and the defined relations: -/
+lemma not_MultisetRedLt_0 [DecidableEq α] [LT α] (M: Multiset α) : ¬ MultisetRedLt M 0 := by
+  intro h
+  cases h with
+  | RedLt X Y a M nonsense _ =>
+    have contra : a ∈ (0 : Multiset α):= by
+      rw [nonsense]
+      simp_all only [Multiset.mem_add, Multiset.mem_singleton, or_true]
+    contradiction
+
+lemma mul_singleton_erase [DecidableEq α] {a : α} {M : Multiset α} :
```

Comment:
Please move to an earlier file (and double check it doesn't already exist).

Reply:
Maybe this can be replaced by using [`Multiset.sub_eq_fold_erase`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Multiset/Basic.html#Multiset.sub_eq_fold_erase) instead?